[WISL] bi-abduction + compiler fixes

ggwg opened the pull request:
## In this PR
Add invalid pointer check on property lookup for WISL compiler. (As discussed in meeting with Sacha on Feb 6)


## Explanation / Evaluation
[Google Doc](https://docs.google.com/document/d/17FIDQDuN9w3AtevHeE-8iUt2sO3t7G37EM9MazrFYco/edit?usp=sharing)

## How to generate specs:

2 test cases are located under `wisl/examples/biabduction/`
```
esy x wisl act wisl/examples/biabduction/read.wisl -o act.gil --specs-to-stdout
```
```
esy x wisl act wisl/examples/biabduction/write.wisl -o act.gil --specs-to-stdout
```
<table>
<tr>
<th> Code </th>
<th> Terminal Output </th>
</tr>
<tr>
<td>

<img width="254" alt="image" src="https://user-images.githubusercontent.com/60797011/225058411-ab6f359b-3f69-4708-88d2-3de1d860b614.png">

</td>
<td>

<img width="622" alt="image" src="https://user-images.githubusercontent.com/60797011/225058333-1b705646-a7c9-4f68-9a89-4bf3554e6097.png">

</td>
</tr>
</table>


## Compiler Change Example

Program Under Test
```
 function llen(x) {
    if (x = null) {
        n := 0
    } else {
        t := [x+1];
        n := llen(t);
        n := n + 1
    };
    return n
}
```

<table>
<tr>
<th> Before </th>
<th> New Change </th>
</tr>
<tr>
<td>

```
proc llen(x) {
          goto [(x = null)] then0 else0;
  then0:  n := 0i;
          goto endif0;
  else0:  gvar0 := "i__add"(x, 1i);
          gvar1 := [getcell](l-nth(gvar0, 0i), l-nth(gvar0, 1i));
          t := l-nth(gvar1, 2i);
          n := "llen"(t);
          gvar2 := "i__add"(n, 1i);
          n := gvar2;
  endif0: skip;
          ret := n;
          return
};
```

</td>
<td>

```
proc llen(x) {
             goto [(x = null)] then0 else0;
  then0:     n := 0i;
             goto endif0;
  else0:     gvar0 := "i__add"(x, 1i);
             goto [((typeOf gvar0) = List)] fail0 continue0;
  fail0:     fail [InvalidPointer]();
  continue0: gvar1 := [getcell](l-nth(gvar0, 0i), l-nth(gvar0, 1i));
             t := l-nth(gvar1, 2i);
             n := "llen"(t);
             gvar2 := "i__add"(n, 1i);
             n := gvar2;
  endif0:    skip;
             ret := n;
             return
};

```

</td>
</tr>
</table>

The new llen procedure should check the type of the input argument and fail with an InvalidPointer error if it is not a list. This is done by adding a type check to the beginning of the procedure. If the type check passes, then the procedure can proceed as normal.

ggwg commented:
**Q**: Whats the standard procedure to test that this compiler change doesn't break anything?

ggwg commented on wisl/lib/ParserAndCompiler/wisl2Gil.ml:
```diff
@@ -724,14 +724,30 @@ let rec compile_stmt_list ?(fname = "main") ?(is_loop_prefix = false) stmtl =
+      let cmd =
+        [
+          (get_annot, None, checkptrcmd);
+          (getval_annot, Some faillab, failcmd);
```
**Q**: What should value of annotation be here?

giltho replied:
Actually, this might be a question for @NatKarmios now. He knows the wisl lifter better than I do.

@ggwg, annotations are used by the debugger to figure out what GIL code corresponds to what WISL code (or other TL code for other instantations).

NatKarmios replied:
Yes, I should have said, Gavin and I talked through this last week, including what the annots should be here, though I'm not sure if he's committed that change.

ggwg replied:
Annotation changes have been updated now. Thanks Nat

NatKarmios commented:
> **Q**: Whats the standard procedure to test that this compiler change doesn't break anything?

Obviously it should compile (`esy`), and then `esy test` will run a wisl verification (as well as tests for other languages, you don't need to worry about them) to check it all works out. This provides a reasonable degree of confidence, though it'd also be a good idea for you to verify a program that uses whatever compilation has been altered (we can talk about this tomorrow).

`esy test` is run automatically by the build job, and I see that it's passing here.

ggwg commented on wisl/examples/act.wisl:
```diff
@@ -0,0 +1,16 @@
+
```
Note: Testing file - to delete

giltho commented:
Just realised @NatKarmios, how is this change going to affect the debugger?

NatKarmios commented:
As long as the annots are correct, it should all work fine. I'll happily test it once this is ready for review.

giltho commented:
We're ready to merge that aren't we?

ggwg commented:
Ready for review. Changes since last version:

- Merge with master
- Update PR description, include Google document
- Clean code by removing all unnecessary comments

ggwg commented:
Not sure why CI / gillian_c_tests (macos-latest) tests are failing.
Only WISL-specific files have been modified so far.
